File: tools/claude/config/skills/scan-git-for-tils/scan_git.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = ["notion-client"]
# ///
"""
Scan GitHub commit history for TIL-worthy commits.

Usage:
    python3 scan_git.py [days]

Arguments:
    days: Number of days to look back (default: 30)

Output:
    JSON with commits for Claude to evaluate

Requires:
    - gh CLI installed and authenticated
    - op CLI installed and authenticated (1Password)
    - uv (for dependency management)
"""

import sys
import json

from git.commits import get_github_username, get_commits
from git.formatting import should_skip_commit, format_markdown
from notion.client import get_assessed_commits_from_notion


def main():
    # Parse arguments
    days = 30
    if len(sys.argv) > 1:
        try:
            days = int(sys.argv[1])
        except ValueError:
            pass

    # Fetch assessed commits from Notion
    assessed_hashes = get_assessed_commits_from_notion()

    # Get GitHub username
    username = get_github_username()
    if not username:
        print(json.dumps({
            "error": "Could not get GitHub username. Is `gh` authenticated?",
            "markdown": "",
            "new_commits": []
        }))
        sys.exit(1)

    # Get commits
    commits = get_commits(days, username)
    total_count = len(commits)

    if not commits:
        print(json.dumps({
            "markdown": format_markdown([], days, 0, 0),
            "new_commits": []
        }))
        sys.exit(0)

    # Filter out already assessed commits and skippable commits
    new_commits = [
        c for c in commits
        if c["full_hash"] not in assessed_hashes and not should_skip_commit(c)
    ]
    new_count = len(new_commits)

    # Prepare output - all commits for Claude to evaluate
    output = {
        "markdown": format_markdown(new_commits, days, new_count, total_count),
        "new_commits": [
            {
                "hash": c["full_hash"],
                "message": c["subject"],
                "repo": c["repo"],
                "date": c["iso_date"]
            }
            for c in new_commits
        ]
    }

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()

File: tools/claude/config/skills/scanning-git-for-tils/scan_git.ts
#!/usr/bin/env bun
/**
 * Scan GitHub commit history for TIL-worthy commits - BUN VERSION
 *
 * Usage: bun run scan_git.bun.ts [days]
 *
 * Demonstrates Bun advantages:
 * - Inline npm dependencies (auto-install on first run)
 * - No deno.json needed
 * - No permission flags needed
 * - TypeScript type safety + Zod validation
 */

import { z } from "zod";
import { getCommits, getGitHubUsername } from "./git/commits.ts";
import { formatMarkdown, shouldSkipCommit } from "./git/formatting.ts";
import { getAssessedCommitsFromNotion } from "./notion/commits.ts";

const CommitSummarySchema = z.object({
  hash: z.string(),
  message: z.string(),
  repo: z.string(),
  date: z.string(),
});

const ScanGitOutputSchema = z.object({
  markdown: z.string(),
  new_commits: z.array(CommitSummarySchema),
});

type ScanGitOutput = z.infer<typeof ScanGitOutputSchema>;

async function main() {
  // Parse arguments
  const days = Bun.argv[2] ? parseInt(Bun.argv[2], 10) || 30 : 30;

  // Fetch assessed commits from Notion
  const assessedHashes = await getAssessedCommitsFromNotion();

  // Get GitHub username
  const username = await getGitHubUsername();
  if (!username) {
    console.log(
      JSON.stringify({
        error: "Could not get GitHub username. Is `gh` authenticated?",
        markdown: "",
        new_commits: [],
      }),
    );
    process.exit(1);
  }

  // Get commits
  const commits = await getCommits(days, username);
  const totalCount = commits.length;

  if (commits.length === 0) {
    const output: ScanGitOutput = {
      markdown: formatMarkdown([], days, 0, 0),
      new_commits: [],
    };
    console.log(JSON.stringify(output));
    process.exit(0);
  }

  // Filter out already assessed commits and skippable commits
  const newCommits = commits.filter((c) =>
    !assessedHashes.has(c.full_hash) && !shouldSkipCommit(c)
  );
  const newCount = newCommits.length;

  // Prepare output
  const output: ScanGitOutput = {
    markdown: formatMarkdown(newCommits, days, newCount, totalCount),
    new_commits: newCommits.map((c) => ({
      hash: c.full_hash,
      message: c.subject,
      repo: c.repo,
      date: c.iso_date,
    })),
  };

  console.log(JSON.stringify(output, null, 2));
}

main();
